SPIKE.md

```diff
@@ -0,0 +1,95 @@
+LoopBack/Juggler except only in one location - the model file.
+
+## Approaches
+
+### 1. Top level ObjectID
+
+A top level `dataType: 'ObjectID'` was suggested, but it was discarded for the following reason:
+
+1. If a top level `ObjectID` is supported for MongoDB, then database-specific types like `MULTIPOLYGON`, `VARCHAR`, `NCLOB` etc
+may also be required to be supported. Considering the number of database-specific types and the number of databases we support,
+we will end up with a really huge list.
+2. Inflexible when switching connectors. Cannot do this:
+
+```js
+{
+  type: 'number',
+  jsonSchema: {
+    format: 'int32',
+  },
+  mysql: {
+    dataType: 'bit',
+  },
+  postgresql: {
+    dataType: 'bit',
+    length: 1,
+  }
+}
+```
+
+### 2. mongodb: {dataType: 'objectID'}
+
+Setting a property to `mongodb: {dataType: 'ObjectID'}`, as it is already being done, is the better approach for marking the 
+property as an `ObjectID`.
+
+## Developer Notes
+
+- The tests in `test/new-objectid.test.js` is a demonstration of the behavior of the `mongodb: {dataType: 'ObjectID'}` property.
+- The state of the result object is determined by both `loopback-connector-mongodb/lib/mongodb.js` and `loopback-datasource-juggler/lib/dao.js`, so changes are required in both files.
+- Changes in `loopback-connector-mongodb/lib/mongodb.js`:
+  - All helper functions and properties like `typeIsObjectId()`, `strictObjectIDCoercion` etc., related to previous behavior of `ObjectId` should be removed or refactored.
+  - Parts of the existing codebase does a good job of coercing string to `ObjectID` when querying the server.
+  - Work is mostly required in coercing `ObjectID` back to string in database result. This is not a simple task. The object that's returned to the user is not generated by a standard pipeline and there objects can behave very differently depending on the query and undergo further modification in `loopback-datasource-juggler/lib/dao.js`.
+- Changes in `loopback-datasource-juggler/lib/dao.js`:
+  - Coercion has to be applied in DAO for some queries.
+  - Depending on the query, there may be different restrictions and rules.
+
+In the time assigned for this spike I could find out the behavior of all the query methods. Also, I didn't have enough time to explore in deep how LB4 handles `mongodb: {dataType: 'ObjectID'}`. I experimented with a simple app with the changes in `loopback-connector-mongodb/lib/mongodb.js` and `loopback-datasource-juggler/lib/dao.js` but was hit with `Type 'MongoDataSource' has no properties in common with type 'LifeCycleObserver'.`; I wonder why it would happen.
+
+## Implementation proposal
+
+- Database-specific connectors should provide their coercion function to DAO.
+- DAO should apply the coercion function on the result object and return it to the user without changing the Model's actual properties.
```

Comment:
Have you considered this simpler implementation?

- In code that's handling property initialization (see [ModelBaseClass.prototype._initProperties](https://github.com/strongloop/loopback-datasource-juggler/blob/261dd1c865e8b197882a161b909b90b0e6c3c2a3/lib/model.js#L202-L207)), we can covert property values provided by the caller to the types defined in model definition.
- ObjectID types are defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`, therefore they will be converted to a string (as we want).
- No extra connector-specific coercion rules are required.

This was one of the reasons why I choose the design where the property `type` is a well-known primitive like a `string` or a `number`, and database-specific information (e.g. how to represent/store the property in the database) is nested under connector-specific config. As long as the database-specific value types play well with JavaScript's built-in coercion (to a string, to a number), no extra work is needed in juggler to support these types.

Consider the "bit" example you described earlier (`{type: 'number', mysql: {dataType: 'bit'}}`). Even if the mysql connector (or driver) returns the property value as a boolean, the generic coercion algorithm will attempt to convert the value to a number, which works exactly as we want.

```
$ node
> +true
1
> +false
0
```

Having said that, there is also the argument that we want to hide the complexity of database-specific types like `ObjectID` from juggler, keep it contained inside the connector only. In that case, all connectors methods should accept stringified object ids and also return stringified object ids. The task of processing model definitions and applying any necessary coercion becomes the responsibility of the connector. 

I believe we have this approach already in place, to support `decimal128` datatype in MongoDB connector.

This solution has an advantage that very little (if any) changes in juggler are necessary, thus we are not going to break existing users of other connectors. It will also make our life easier in the future, when we implement a LB4 `Repository` class that's calling connectors directly, removing juggler from the picture.

Personally, I find this second approach better for the longer term.

Either way, please update your proposal to take into account the existing code for `decimal128` and come up with a consistent solution for both `ObjectID` and `decimal128` types. IMO, these types should be handled the same way.